rfcs/framework-0063-notification-system.md

```diff
@@ -0,0 +1,422 @@
+        ret = []
+        if self.include_creator:
+            ret.append(request.created_by.resolve())
+        if self.include_receiver:
+            ret.append(request.receiver.resolve())
+        return ret
+
+# invenio_communities/notifications.py
+class CommunityRecipients:
+    def __init__(self, roles=None):
+        self.roles = roles
+
+    def __call___(self, notification, recipients: list):
+        ret = []
+        for rec in recipients:
+            if isinstance(rec, Community):
+                comm = rec
+                members = Member.get_members(comm.id)
+                for m in members:
+                    if not m.user_id:
+                        continue
+                    if self.roles and m.role not in self.roles:
+                        continue
+                    user = m.relations.user.dereference()
+                    notif_pref = user.preferences["notifications"]
+                    com_pref = notif_pref.get(comm.id)
+                    if com_pref:
+                        # check if notification is enabled/disabled for the member
+                        # if ...:
+                        #     continue
+                    ret.append(user)
+            else:
+                ret.append(rec)
+        return ret
+
+# invenio_users_resources/notifications.py
+class UserRecipient:
+    def __call__(self, notification, recipients: list)
+        ret = []
+        for rec in recipients:
+            if isinstance(rec, User):
+                if rec.preferences["notifications"]["enabled"]
+                    ret.append(rec.dump())
+            else:
+                ret.append(rec)
+        return ret
+
+#
+# Backends
+#
+# invenio_users_resources/notifications.py
+class UserEmailBackend:
+    def __call__(self, notification, recipients: list)
+        for rec in recipients:
+            user = rec.data
+            rec.backends.append({
+                "backend": "email",
+                "to": f"{user.profile.full_name} <{user.email}>",
+            })
+
+#
+# Usage in a service method
+#
+uow.register(NotificationOp(CommunityRecordInclusionNotificationBuilder.build(
+    {
+        "request": request,
+        "community": community,
+        "record": record,
+    }
+)))
+```
+
+#### Notification examples
+
+| Action             | Who triggers   | Who receives                                                                            |
+| ------------------ | -------------- | --------------------------------------------------------------------------------------- |
+| Comment create     | Creating user  | All involved in conversation                                                            |
+| Comment edit       | Editing user   | All involved in conversation                                                            |
+| Comment delete     | Deleting user  | All involved in conversation                                                            |
```

Comment:
Comment edit/delete might be overkill. I don't think that GitHub sends out notifications when a comment is updated or deleted.